src/gadgets/hash/blake3.rs

```diff
@@ -0,0 +1,652 @@
+    b: usize,
+    c: usize,
+    d: usize,
+    mx: U32,
+    my: U32,
+) {
+    let tmp0 = U32::wrapping_add(circuit, state[a], state[b]);
+    state[a] = U32::wrapping_add(circuit, tmp0, mx);
+    state[d] = U32::rotate_right(U32::xor(circuit, state[d], state[a]), 16);
+    state[c] = U32::wrapping_add(circuit, state[c], state[d]);
+    state[b] = U32::rotate_right(U32::xor(circuit, state[b], state[c]), 12);
+
+    let tmp0 = U32::wrapping_add(circuit, state[a], state[b]);
+    state[a] = U32::wrapping_add(circuit, tmp0, my);
+    state[d] = U32::rotate_right(U32::xor(circuit, state[d], state[a]), 8);
+    state[c] = U32::wrapping_add(circuit, state[c], state[d]);
+    state[b] = U32::rotate_right(U32::xor(circuit, state[b], state[c]), 7);
+}
+
+fn round<C: CircuitContext>(circuit: &mut C, state: &mut [U32; 16], m: &[U32; 16]) {
+    // Mix the columns.
+    g(circuit, state, 0, 4, 8, 12, m[0], m[1]);
+    g(circuit, state, 1, 5, 9, 13, m[2], m[3]);
+    g(circuit, state, 2, 6, 10, 14, m[4], m[5]);
+    g(circuit, state, 3, 7, 11, 15, m[6], m[7]);
+    // Mix the diagonals.
+    g(circuit, state, 0, 5, 10, 15, m[8], m[9]);
+    g(circuit, state, 1, 6, 11, 12, m[10], m[11]);
+    g(circuit, state, 2, 7, 8, 13, m[12], m[13]);
+    g(circuit, state, 3, 4, 9, 14, m[14], m[15]);
+}
+
+fn permute(m: &mut [U32; 16]) {
+    let mut permuted = [U32([FALSE_WIRE; 32]); 16];
+    for i in 0..16 {
+        permuted[i] = m[MSG_PERMUTATION[i] as usize];
+    }
+    *m = permuted;
+}
+
+fn compress<C: CircuitContext>(
+    circuit: &mut C,
+    chaining_value: &[U32; 8],
+    block_words: &[U32; 16],
+    counter: u64,
+    block_len: U32,
+    flags: U32,
+) -> [U32; 16] {
+    let counter_low = U32::from_constant(counter as u32);
+    let counter_high = U32::from_constant((counter >> 32) as u32);
+    #[rustfmt::skip]
+    let iv: [U32; 8] = get_iv();
+    let mut state = [
+        chaining_value[0],
+        chaining_value[1],
+        chaining_value[2],
+        chaining_value[3],
+        chaining_value[4],
+        chaining_value[5],
+        chaining_value[6],
+        chaining_value[7],
+        iv[0],
+        iv[1],
+        iv[2],
+        iv[3],
+        counter_low,
+        counter_high,
+        block_len,
+        flags,
+    ];
+
+    let mut block = *block_words;
+
+    round(circuit, &mut state, &block); // round 1
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 2
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 3
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 4
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 5
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 6
+    permute(&mut block);
+    round(circuit, &mut state, &block); // round 7
+
+    for i in 0..8 {
+        state[i] = U32::xor(circuit, state[i], state[i + 8]);
+        state[i + 8] = U32::xor(circuit, state[i + 8], chaining_value[i]);
```

Comment:
not technically needed if we only use 256 bits - we can return `BitIntWires` or `[U32; 8]`